File: c_programs_and_containers/build_blocks/knowledge_base/sqlite3/data_structures/python/bit_mask_operations.py
"""
SQLite3 Bit Mask Operations Class
Manages bit masks and flag registers for distributed node control systems
"""

import json
import sqlite3
from typing import Dict, Optional, Any


class BitMaskOperations:
    """
    Manages bit mask operations for distributed control system nodes.
    Provides methods to create, read, and update bit masks and flag dictionaries
    stored in SQLite3.
    """
    
    def __init__(self, conn: sqlite3.Connection, bit_mask_table_name: str = "bit_mask_table"):
        """
        Initialize the bit mask operations handler.
        Creates the table if it doesn't exist.
        
        Args:
            conn: Active SQLite3 connection object
            bit_mask_table_name: Name of the table to use for bit mask storage
        """
        self.conn = conn
        self.table_name = bit_mask_table_name
        
    
    def create_table(self):
        """
        Creates a fresh bit mask table, dropping any existing table with the same name.
        Table schema:
            - node_id: TEXT PRIMARY KEY (unique identifier for node)
            - bit_mask: INTEGER (64-bit integer mask)
        
        """
        # Drop existing table if it exists
        drop_table_query = f"DROP TABLE IF EXISTS {self.table_name}"
        
        # Create fresh table
        create_table_query = f"""
            CREATE TABLE {self.table_name} (
                node_id TEXT PRIMARY KEY,
                bit_mask INTEGER NOT NULL DEFAULT 0
            )
        """
        
        cursor = self.conn.cursor()
        cursor.execute(drop_table_query)
        cursor.execute(create_table_query)
        self.conn.commit()
    
    def create_entry(self, node_id: str, bit_mask: int = 0) -> bool:
        """
        Create a new entry in the bit mask table.
        
        Args:
            node_id: Unique identifier for the node
            bit_mask: Initial bit mask value (default: 0)
        
        Returns:
            True if entry created successfully, False otherwise
        
        Raises:
            sqlite3.IntegrityError: If node_id already exists
        """
        insert_query = f"""
            INSERT INTO {self.table_name} (node_id, bit_mask)
            VALUES (?, ?)
        """
        
        try:
            cursor = self.conn.cursor()
            cursor.execute(insert_query, (node_id, bit_mask))
            self.conn.commit()
            return True
        except sqlite3.IntegrityError as e:
            self.conn.rollback()
            raise sqlite3.IntegrityError(
                f"Node ID '{node_id}' already exists in table '{self.table_name}'"
            ) from e
        except Exception as e:
            self.conn.rollback()
            raise
    
    def get_bit_mask(self, node_id: str) -> Optional[int]:
        """
        Retrieve the bit mask value for a given node.
        
        Args:
            node_id: Unique identifier for the node
        
        Returns:
            64-bit integer bit mask value, or None if node not found
        """
        select_query = f"""
            SELECT bit_mask FROM {self.table_name}
            WHERE node_id = ?
        """
        
        cursor = self.conn.cursor()
        cursor.execute(select_query, (node_id,))
        result = cursor.fetchone()
        
        if result is None:
            return None
        
        return result[0]
    
    def set_bit_mask(self, node_id: str, new_bits: int, change_mask: int = -1) -> bool:
        """
        Atomically update specific bits in the bit_mask for a given node.

        This performs the following SQL-level operation:
        new_mask = (current_mask & (~change_mask)) | (new_bits & change_mask)
        
        Args:
            node_id: Unique identifier for the node
            new_bits: The integer containing the new bit values to apply.
            change_mask: A 64-bit mask where each '1' indicates
                        a bit to be updated using the value from new_bits.
                        Defaults to -1 (all 64 bits, 0xFFF...FFF)
                        for a full overwrite.
        
        Returns:
            True if the update was successful (row affected), False otherwise.
        
        Raises:
            ValueError: If new_bits or change_mask is out of the
                        valid 64-bit signed integer range.
        """
        # Define 64-bit signed integer limits
        SIGNED_64_BIT_MIN = -9223372036854775808
        SIGNED_64_BIT_MAX = 9223372036854775807
        
        # Validate both inputs are within the 64-bit signed integer range
        if not (SIGNED_64_BIT_MIN <= new_bits <= SIGNED_64_BIT_MAX):
            raise ValueError(f"new_bits must be a valid 64-bit integer, got {new_bits}")
            
        if not (SIGNED_64_BIT_MIN <= change_mask <= SIGNED_64_BIT_MAX):
            raise ValueError(f"change_mask must be a valid 64-bit integer, got {change_mask}")

        # This single SQL query performs the read, modify, and write atomically.
        # 1. (bit_mask & (~?)) -> Clears the bits we want to change.
        # 2. (? & ?) -> Isolates the new bits to be set.
        # 3. (... | ...) -> Combines the two, updating the mask.
        update_query = f"""
            UPDATE {self.table_name}
            SET bit_mask = (bit_mask & (~?)) | (? & ?)
            WHERE node_id = ?
        """
        
        # The parameters are: (change_mask, new_bits, change_mask, node_id)
        params = (change_mask, new_bits, change_mask, node_id)
        
        rows_affected = 0
        try:
            cursor = self.conn.cursor()
            cursor.execute(update_query, params)
            rows_affected = cursor.rowcount
            self.conn.commit()
        except Exception:
            # Rollback in case of any SQL error
            self.conn.rollback()
            raise  # Re-raise the exception after rollback
            
        return rows_affected > 0

    
    def get_entry(self, node_id: str) -> Optional[Dict[str, Any]]:
        """
        Retrieve complete entry for a given node.
        
        Args:
            node_id: Unique identifier for the node
        
        Returns:
            Dictionary with keys: node_id, bit_mask
            Returns None if node not found
        """
        select_query = f"""
            SELECT node_id, bit_mask
            FROM {self.table_name}
            WHERE node_id = ?
        """
        
        cursor = self.conn.cursor()
        cursor.execute(select_query, (node_id,))
        result = cursor.fetchone()
        
        if result is None:
            return None
        
        return {
            'node_id': result[0],
            'bit_mask': result[1],
        }
    
    def delete_entry(self, node_id: str) -> bool:
        """
        Delete an entry from the bit mask table.
        
        Args:
            node_id: Unique identifier for the node
        
        Returns:
            True if entry deleted, False if node not found
        """
        delete_query = f"""
            DELETE FROM {self.table_name}
            WHERE node_id = ?
        """
        
        cursor = self.conn.cursor()
        cursor.execute(delete_query, (node_id,))
        rows_affected = cursor.rowcount
        
        self.conn.commit()
        return rows_affected > 0
    
    '''
    Dictionary is not allowed to be modified after creation
    def update_flag_dictionary(self, node_id: str, flag_dictionary: Dict[str, Any]) -> bool:
        """
        Update the flag dictionary for a given node.
        
        Args:
            node_id: Unique identifier for the node
            flag_dictionary: New flag dictionary to store
        
        Returns:
            True if update successful, False if node not found
        """
        update_query = f"""
            UPDATE {self.table_name}
            SET flag_regist = ?
            WHERE node_id = ?
        """
        
        cursor = self.conn.cursor()
        cursor.execute(update_query, (json.dumps(flag_dictionary), node_id))
        rows_affected = cursor.rowcount
        
        self.conn.commit()
        return rows_affected > 0
    '''
    
    def list_all_nodes(self) -> list:
        """
        List all node IDs in the table.
        
        Returns:
            List of all node_id strings
        """
        select_query = f"""
            SELECT node_id FROM {self.table_name}
            ORDER BY node_id
        """
        
        cursor = self.conn.cursor()
        cursor.execute(select_query)
        return [row[0] for row in cursor.fetchall()]

File: luajit_programs_and_containers/nano_data_center_base/commissioning_software/kb/postgres/data_structures/bit_mask_operations.lua
--[[
  BitMaskOperations - Manages bit masks and flag registers for distributed
  node control systems.  LuaJIT port using luadbi-postgresql.

  Usage:
    local BitMaskOperations = require("bit_mask_operations")
    local bm = BitMaskOperations.new(dbh, "bit_mask_table")
    bm:create_table()
    bm:create_entry("node_1", 0)
    bm:set_bit_mask("node_1", 0xFF, -1)
]]

local bit = require("bit")   -- LuaJIT bit library

local BitMaskOperations = {}
BitMaskOperations.__index = BitMaskOperations

---------------------------------------------------------------------------
-- Helpers
---------------------------------------------------------------------------

local function esc(val)
  if val == nil then return "NULL" end
  local s = tostring(val)
  s = s:gsub("'", "''")
  return "'" .. s .. "'"
end

local function raw_query(dbh, sql_str)
  local sth, err = dbh:prepare(sql_str)
  if not sth then error("Prepare error: " .. tostring(err) .. "\nSQL: " .. sql_str) end
  local ok, exec_err = sth:execute()
  if not ok then error("Execute error: " .. tostring(exec_err) .. "\nSQL: " .. sql_str) end
  return sth
end

local function fetch_all(dbh, sql_str)
  local sth = raw_query(dbh, sql_str)
  local rows = {}
  while true do
    local row = sth:fetch(true)
    if not row then break end
    local copy = {}
    for k, v in pairs(row) do copy[k] = v end
    rows[#rows + 1] = copy
  end
  return rows
end

local function fetch_one(dbh, sql_str)
  local sth = raw_query(dbh, sql_str)
  local row = sth:fetch(true)
  if not row then return nil end
  local copy = {}
  for k, v in pairs(row) do copy[k] = v end
  return copy
end

local function exec_dml(dbh, sql_str)
  local sth = raw_query(dbh, sql_str)
  local affected = sth:affected()
  return affected or 0
end

---------------------------------------------------------------------------
-- Constructor
---------------------------------------------------------------------------

--- Create a new BitMaskOperations instance.
-- @param dbh  DBI connection handle
-- @param table_name  table name (default "bit_mask_table")
function BitMaskOperations.new(dbh, table_name)
  local self = setmetatable({}, BitMaskOperations)
  self.dbh        = dbh
  self.table_name = table_name or "bit_mask_table"
  return self
end

---------------------------------------------------------------------------
-- Table management
---------------------------------------------------------------------------

function BitMaskOperations:create_table()
  raw_query(self.dbh, string.format("DROP TABLE IF EXISTS %s", self.table_name))
  raw_query(self.dbh, string.format([[
    CREATE TABLE %s (
      node_id  VARCHAR(255) PRIMARY KEY,
      bit_mask BIGINT NOT NULL DEFAULT 0
    )
  ]], self.table_name))
  self.dbh:commit()
end

---------------------------------------------------------------------------
-- CRUD
---------------------------------------------------------------------------

--- Create a new entry.
-- @return true on success
function BitMaskOperations:create_entry(node_id, bit_mask_val)
  bit_mask_val = bit_mask_val or 0
  local sql_str = string.format(
    "INSERT INTO %s (node_id, bit_mask) VALUES (%s, %s)",
    self.table_name, esc(node_id), tostring(bit_mask_val))

  local ok, err = pcall(function()
    exec_dml(self.dbh, sql_str)
    self.dbh:commit()
  end)
  if not ok then
    pcall(function() self.dbh:rollback() end)
    error(string.format("Node ID '%s' may already exist: %s", node_id, tostring(err)))
  end
  return true
end

--- Retrieve bit_mask for a node.
-- @return integer or nil
function BitMaskOperations:get_bit_mask(node_id)
  local sql_str = string.format(
    "SELECT bit_mask FROM %s WHERE node_id = %s",
    self.table_name, esc(node_id))
  local row = fetch_one(self.dbh, sql_str)
  if not row then return nil end
  return tonumber(row.bit_mask)
end

--- Atomically update specific bits:
--   new_mask = (current & ~change_mask) | (new_bits & change_mask)
-- @param node_id   string
-- @param new_bits  integer (64-bit)
-- @param change_mask integer (default -1 = all bits)
-- @return true if row updated
function BitMaskOperations:set_bit_mask(node_id, new_bits, change_mask)
  change_mask = change_mask or -1

  local sql_str = string.format([[
    UPDATE %s
       SET bit_mask = (bit_mask & (~(%s)::bigint)) | ((%s)::bigint & (%s)::bigint)
     WHERE node_id = %s
  ]], self.table_name,
     tostring(change_mask), tostring(new_bits), tostring(change_mask),
     esc(node_id))

  local ok, err = pcall(function()
    local affected = exec_dml(self.dbh, sql_str)
    self.dbh:commit()
    return affected
  end)

  if not ok then
    pcall(function() self.dbh:rollback() end)
    error(tostring(err))
  end
  -- err here is actually the return value from the pcall success path
  return (err or 0) > 0
end

--- Retrieve full entry as dict.
function BitMaskOperations:get_entry(node_id)
  local sql_str = string.format(
    "SELECT node_id, bit_mask FROM %s WHERE node_id = %s",
    self.table_name, esc(node_id))
  local row = fetch_one(self.dbh, sql_str)
  if not row then return nil end
  row.bit_mask = tonumber(row.bit_mask)
  return row
end

--- Delete an entry.
-- @return true if deleted
function BitMaskOperations:delete_entry(node_id)
  local sql_str = string.format(
    "DELETE FROM %s WHERE node_id = %s",
    self.table_name, esc(node_id))
  local affected = exec_dml(self.dbh, sql_str)
  self.dbh:commit()
  return affected > 0
end

--- List all node IDs.
function BitMaskOperations:list_all_nodes()
  local sql_str = string.format(
    "SELECT node_id FROM %s ORDER BY node_id", self.table_name)
  local rows = fetch_all(self.dbh, sql_str)
  local rv = {}
  for _, r in ipairs(rows) do rv[#rv + 1] = r.node_id end
  return rv
end

return BitMaskOperations